File: src/spice2x/api/resources/python/spiceapi/buttons.py
from .connection import Connection
from .request import Request


def buttons_read(con: Connection):
    res = con.request(Request("buttons", "read"))
    return res.get_data()


def buttons_write(con: Connection, button_state_list):
    req = Request("buttons", "write")
    for state in button_state_list:
        req.add_param(state)
    con.request(req)


def buttons_write_reset(con: Connection, button_names=None):
    req = Request("buttons", "write_reset")

    # reset all buttons
    if not button_names:
        con.request(req)
        return

    # reset specified buttons
    for button_name in button_names:
        req.add_param(button_name)
    con.request(req)

File: src/spice2x/api/resources/dart/spiceapi/src/wrappers/buttons.dart
part of spiceapi;

class ButtonState {
  String name;
  double state;
  bool active;

  ButtonState(this.name, this.state);
  ButtonState._fromRead(this.name, this.state, this.active);
}

Future<List<ButtonState>> buttonsRead(Connection con) {
  var req = Request("buttons", "read");
  return con.request(req).then((res) {

    // build states list
    List<ButtonState> states = [];
    for (List state in res.getData()) {
      states.add(ButtonState._fromRead(
        state[0],
        state[1],
        state[2],
      ));
    }

    // return it
    return states;
  });
}

Future<void> buttonsWrite(Connection con, List<ButtonState> states) {
  var req = Request("buttons", "write");

  // add params
  for (var state in states) {
    var obj = [
      state.name,
      state.state
    ];
    req.addParam(obj);
  }

  return con.request(req);
}

Future<void> buttonsWriteReset(Connection con, List<String> names) {
  var req = Request("buttons", "write_reset");

  // add params
  for (var name in names)
    req.addParam(name);

  return con.request(req);
}
